packages/framework/agent-scheduler/src/scheduler.ts

```diff
@@ -257,7 +257,14 @@ export class AgentScheduler extends TypedEventEmitter<IAgentSchedulerEvents> imp
             if (this.isActive() && currentClient === this.clientId) {
                 this.onNewTaskAssigned(key);
             } else {
-                await this.onTaskReassigned(key, currentClient);
+                // The call below mutates the consensusRegisterCollection in
+                // its event handler, which is not safe.
+                // We need to force this to be part of a different batch of ops by
+                // scheduling a microtask in order to work around the current validations.
+                // This is not recommended and should be avoided.
+                await Promise.resolve().then(async () => {
+                    await this.onTaskReassigned(key, currentClient);
```

Comment:
It's probably safe (by looking inside of onTaskReassigned), but not sure. In general, converting synch flow to async is not safe as state changed underneath you while waiting for the promise to be scheduled.

Reply:
I'm also a bit weary of this approach as it the state of the world can greatly be altered in between batches. But the code was already async, right?

Do you have an idea of a different approach to this? I mean it would also be great to have a consistent migration strategy which we can present to our customers in case they're currently using this pattern.

Alternatively, we can do something like double-checking: check condition and schedule the next batch -> before running the op generating code in the next batch, check the condition again and abandon if the condition is no longer met.